### PR TITLE
Further optimize fm.c.

### PR DESCRIPTION
**Pull request by movrsi**
Branching has been reduced in this commit. This commit deals mostly with branching optimizations. 

Im pretty sure this cover all the routines that are called upon most frequent, if not let me know and i can optimize other routines within fm.c if this isn't the case.

### CODE REVIEW

**Comment by nukeykt** [comment]
idk. this one seems to hurt code readability too much for a little speed gain

**Comment by movrsi** [comment]
Yes that is a problem with some of these optimizations. Depending on which parts seem too much; i could cherry pick them. Tried to keep somethings within inline routines at an attempt to try and keep it readable. Though i'm pretty sure these are most called routines within the fm chip.

**Comment by ogamespec** [comment]
You've gone through the whole module. And most of the edits are things that the compiler is already able to do itself.
To understand where the slowdowns are, you can use a profiler, and then do spot optimization.
Such PR is very difficult to do a review and there is little confidence that nothing will break.

File: fm.c [hunk 1]
```diff
@@ -79,21 +79,19 @@ void FM_Prescaler(fm_prescaler_t *chip)
 
 void FM_HandleIO(fm_t *chip)
 {
-    int write_data = chip->input.cs && chip->input.wr && (chip->input.address & 1) == 1 && !chip->input.ic;
```
Comment (ogamespec) [inline fm.c:82]: no speed gain.

File: fm.c [hunk 2]
```diff
@@ -79,21 +79,19 @@ void FM_Prescaler(fm_prescaler_t *chip)
...
+    //const int io_dir = chip->input.cs && chip->input.rd && !chip->input.ic;
+    //const int data_enable = !io_dir && !chip->input.ic;
 
     if (chip->input.cs && chip->input.wr)
     {
         chip->data_latch = chip->input.data;
         chip->bank_latch = (chip->input.address >> 1) & 1;
     }
 
-    if (write_addr)
```
Comment (ogamespec) [inline fm.c:94]: no speed gain.

File: fm.c [hunk 3]
```diff
@@ -79,21 +79,19 @@ void FM_Prescaler(fm_prescaler_t *chip)
...
+    if ((chip->input.cs && chip->input.wr && (chip->input.address & 1) == 0) || chip->input.ic)
         chip->write_addr_trig = 1;
-    if (write_data)
```
Comment (ogamespec) [inline fm.c:96]: no speed gain.

File: fm.c [hunk 4]
```diff
@@ -102,14 +100,14 @@ void FM_HandleIO(fm_t *chip)
...
-    if (!read_enable)
+    if (!(chip->input.cs && chip->input.rd && !chip->input.ic))
     {
         chip->status_timer_a_dlatch = chip->timer_a_status[1];
         chip->status_timer_b_dlatch = chip->timer_b_status[1];
     }
 }
 
-int FM_GetBus(fm_t *chip)
+int FM_GetBus(const fm_t *chip)
```
Comment (ogamespec) [inline fm.c:110]: no speed gain.

File: fm.c [hunk 5]
```diff
@@ -669,219 +658,642 @@ void FM_HandleIO2(fm_t *chip)
...
+    // pan
+    chip->chan_pan[0][0] &= ~1;
+    chip->chan_pan[1][0] &= ~1;
+
+    // a register
+    chip->chan_a4[0] = 0;
+    chip->chan_ac[0] = 0;
+}
+
+static inline void fm_rotate_registers(fm_t *chip, const int sel, const int from)
```
Comment (ogamespec) [inline fm.c:1030]: What difference does it make when to rotate, here or as it was before. The readability has degraded.

File: fm.c [hunk 6]
```diff
@@ -2467,103 +2891,92 @@ void FM_Accumulator2(fm_t* chip)
...
         chip->timer_b_cnt[0] = time;
-    chip->timer_b_of[0] = (time & 256) != 0;
 
+    chip->timer_b_of[0] = (time & 256) != 0;
     chip->timer_dlatch = chip->fsm_clock_timers;
 }
 
 void FM_Timers2(fm_t *chip)
 {
-    int read_enable = chip->input.cs && chip->input.rd && !chip->input.ic;
```
Comment (ogamespec) [inline fm.c:2566]: The readability has degraded.

**Comment by ogamespec** [comment]
@nukeykt I looked at the PR, except for useless refactoring, I didn't see anything of value.
@movrsi Can you point out places where optimization was really done, not just reworked for (as you think) optimization by throwing out local variables (which the compiler already knows how to do on its own)?

File: fm.c (hunk 5 above)
Comment (movrsi) [reply fm.c:1030]: Previous method of rotation even when unrolled included 90-100 usages of CH_ROTATE and SLOT_ROTATE all of which is branching and cannot be predicted by the compiler. doing it as `fm_swap_registers` and `fm_rotate_registers` takes out the branching completely and uses a lookup table to determine which path to go for either swapping or rotating the values.

File: fm.c (hunk 2 above)
Comment (movrsi) [reply fm.c:94]: This isn't for a gain of speed; it cleans up stack memory as these values could be constant but for the most part are only used once while defined.

File: fm.c (hunk 5 above)
Comment (ogamespec) [reply fm.c:1030]: I agree with you there. How about we redo the edits in small batches?

**Comment by ogamespec** [comment]
Does it make sense to optimize FM at all? I have shown that FM takes about 2% of performance.

image

**Comment by nukeykt** [comment]
I redid most important optimization with less code redundancy in this commit def65d7bbfa1edc4712c963b5353a632c4e4c8de. Thank you for the idea for this one. Other optimizations seems are not as important, so I'm closing PR

**Comment by movrsi** [comment]
Understandable, some of the code provided was to clean up things up a bit. next time ill do pr's in batches to prevent such a mess. ill get into the habit of attempting to use macros to prevent redundant code where possible.

**Pull request closed without merging**